x-pack/plugins/security_solution/server/endpoint/services/lookup_agent.ts

```diff
@@ -0,0 +1,30 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { SearchResponse } from 'elasticsearch';
+import { HostMetadata } from '../../../common/endpoint/types';
+import { SecuritySolutionRequestHandlerContext } from '../../types';
+import { getESQueryHostMetadataByIDs } from '../routes/metadata/query_builders';
+import { EndpointAppContext } from '../types';
+
+export async function GetAgentIDsForEndpoints(
+  endpointIDs: string[],
+  requestHandlerContext: SecuritySolutionRequestHandlerContext,
+  endpointAppContext: EndpointAppContext
+): Promise<string[]> {
+  const queryStrategy = await endpointAppContext.service
+    ?.getMetadataService()
+    ?.queryStrategy(requestHandlerContext.core.savedObjects.client);
+
+  const query = getESQueryHostMetadataByIDs(endpointIDs, queryStrategy!);
```

Comment:
If `queryStrategy` cannot be `null` I think you should check it before that and do something (maybe throw an error?). If it can be null, maybe add it as conditional in `getESQueryHostMetadataByIDs` but I think it must be defined to retrieve the index right?

Reply:
I think the root of all this `? !` is because services off of the `endpointAppContextService` are all `serviceType | undefined` because the plugin starts with them undefined, and goes through and populates them all at start (after setup) time. So there is a brief period during setup that the services are actually undefined, but always exist after that.

There is likely a different way we should specify that these will all be certainly defined after startup, but this is straight copy-paste from how we do it in `getHostData()`: https://github.com/elastic/kibana/blob/master/x-pack/plugins/security_solution/server/endpoint/routes/metadata/handlers.ts#L173-L179